doc/guides/zephyr_cmake_package.rst

```diff
@@ -0,0 +1,368 @@
+In this example ``my_first_app`` is a Zephyr workspace application.
+
+.. note:: The root of a Zephyr workspace is identical to ``west topdir`` if the workspace was
+          installed using ``west``
+
+
+Zephyr freestanding application
+===============================
+
+A Zephyr freestanding application is a Zephyr application located outside of a Zephyr workspace.
+
+
+.. code-block:: none
+
+   <projects>/zephyr-workspace
+   ├── zephyr
+   ├── bootloader
+   └── ...
+
+   <home>/app
+   ├── CMakeLists.txt
+   ├── prj.conf
+   └── src
+       └── main.c
+
+In this example ``app`` is a Zephyr freestanding application.
+
+
+Zephyr Base Environment Setting
+*******************************
+
+The Zephyr CMake package search functionality allows for explicitly specifying
+a Zephyr base using an environment variable.
+
+To do this, use the following ``find_package()`` syntax:
+
+.. code-block:: cmake
+
+   find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
+
+This syntax instructs CMake to first search for Zephyr at the location pointed
+to by the :envvar:`ZEPHYR_BASE` environment variable then use the normal search
+paths.
+
+.. _zephyr_cmake_search_order:
+
+Zephyr CMake Package Search Order
+*********************************
+
+When ``ZEPHYR_BASE`` is not used for searching, the Zephyr installation matching
+the following criteria will be used:
+
+* A Zephyr repository application will use the Zephyr in which it is located.
+  For example:
+
+  .. code-block:: none
+
+        <projects>/zephyr-workspace/zephyr
+        └── samples
+            └── hello_world
+
+  in this example, ``hello_world`` will use ``<projects>/zephyr-workspace/zephyr``.
+
+
+* Zephyr workspace application will use the Zephyr that share the same workspace.
+  For example:
+
+  .. code-block:: none
+
+     <projects>/zephyr-workspace
+     ├── zephyr
+     ├── ...
+     └── my_applications
+          └── my_first_app
+
+  in this example, ``my_first_app`` will use ``<projects>/zephyr-workspace/zephyr`` as this Zephyr
+  is located in the same workspace as the Zephyr workspace application.
+
+
+* Zephyr freestanding application will use the Zephyr registered in the CMake user package registry.
+  For example:
+
+  .. code-block:: none
+
+     <projects>/zephyr-workspace-1
+     └── zephyr                       (Not exported to CMake)
+
+     <projects>/zephyr-workspace-2
+     └── zephyr                       (Exported to CMake)
+
+     <home>/app
+     ├── CMakeLists.txt
+     ├── prj.conf
+     └── src
+         └── main.c
+
+  in this example, only ``<projects>/zephyr-workspace-2/zephyr`` is exported to the CMake package
+  registry and therefore this Zephyr will be used by the Zephyr freestanding application
+  ``<home>/app``.
+
+  If user wants to test the application with ``<projects>/zephyr-workspace-1/zephyr``, this can be
+  done by setting ``ZEPHYR_BASE=<projects>/zephyr-workspace-1/zephyr`` in environment, before
+  running CMake.
+
+
```

Comment:
Add a comment about env `ZEPHYR_BASE` being ignored if rebuilding an existing build that used `find_package()`

Reply:
Good point.
Although that would be what I expect from a build system, but as current Zephyr behaves differently, it is important to highlight.

Reply:
Added:
```
 .. note::

     The Zephyr package selected on the first CMake invocation will be used for all subsequent
     builds. To change the Zephyr package, as example to test the application using environment
     ``ZEPHYR_BASE`` then it is neccessary to do a pristine build first 
     (See :ref:`application_rebuild`). 
```

Reply:
The act of providing ZEPHYR_BASE through HINTS is described in three different ways in the document:

- "`find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`"
- "When ``ZEPHYR_BASE`` is not used for searching"
- "to test the application using environment ``ZEPHYR_BASE``"

It might not be immediately obvious that these are the same thing. Tell me if they aren't, because I'm not 100% sure :P.
To streamline it, maybe it's good to introduce a terminology in the "Zephyr Base Environment Setting" and use it throughout.

Reply:
Did a clean-up on this as you suggested.
Feel free to take a second look.